engine/time/src/main/java/io/deephaven/time/calendar/BusinessCalendar.java

```diff
@@ -1,774 +1,2164 @@
+            }
+
+            date = date.plusDays(1);
+        }
+    }
+
+    private static class YearData {
+        private final Instant start;
+        private final Instant end;
+        private final long businessTimeNanos;
+
+        public YearData(final Instant start, final Instant end, final long businessTimeNanos) {
+            this.start = start;
+            this.end = end;
+            this.businessTimeNanos = businessTimeNanos;
+        }
+    }
+
+    private void populateCachedYearData() {
+        // Only cache complete years, since incomplete years can not be fully computed.
+
+        final int yearStart =
+                firstValidDate.getDayOfYear() == 1 ? firstValidDate.getYear() : firstValidDate.getYear() + 1;
+        final int yearEnd = ((lastValidDate.isLeapYear() && lastValidDate.getDayOfYear() == 366)
+                || lastValidDate.getDayOfYear() == 365) ? lastValidDate.getYear() : lastValidDate.getYear() - 1;
+
+        for (int year = yearStart; year <= yearEnd; year++) {
+            final LocalDate startDate = LocalDate.ofYearDay(year, 1);
+            final LocalDate endDate = LocalDate.ofYearDay(year + 1, 1);
+            final ZonedDateTime start = startDate.atTime(0, 0).atZone(timeZone());
+            final ZonedDateTime end = endDate.atTime(0, 0).atZone(timeZone());
+
+            LocalDate date = startDate;
+            long businessTimeNanos = 0;
+
+            while (date.isBefore(endDate)) {
+                final CalendarDay<Instant> bs = this.calendarDay(date);
+                businessTimeNanos += bs.businessNanos();
+                date = date.plusDays(1);
+            }
+
+            final YearData yd = new YearData(start.toInstant(), end.toInstant(), businessTimeNanos);
+            cachedYearData.put(year, yd);
+        }
+    }
+
+    private YearData getYearData(final int year) {
+        final YearData yd = cachedYearData.get(year);
+
+        if (yd == null) {
+            throw new InvalidDateException("Business calendar does not contain a complete year for: year=" + year);
+        }
+
+        return yd;
+    }
+
+    // endregion
+
+    // region Constructors
+
+    /**
+     * Creates a new business calendar.
+     *
+     * @param name calendar name.
+     * @param description calendar description.
+     * @param timeZone calendar time zone.
+     * @param firstValidDate first valid date for the business calendar.
+     * @param lastValidDate last valid date for the business calendar.
+     * @param standardBusinessDay business day schedule for a standard business day
+     * @param weekendDays weekend days
+     * @param holidays holidays. Business day schedules for all holidays. A holiday is a date that has a schedule that
+     *        is different from the schedule for a standard business day or weekend.
+     * @throws RequirementFailure if any argument is null.
+     */
+    public BusinessCalendar(final String name, final String description, final ZoneId timeZone,
```

Comment:
Is this meant to be user-visible? If we need it to be user visible, we should move it to a static method; regardless, we should make the constructor package-private (to support theoretical interface refactor in the future if necessary).